File: _2021/go/day1/day1.go
package day1

import (
	"fmt"
	"strconv"
	"strings"

	"advent/utils"
)

func parseInput() []int {
	data := utils.LoadPuzzleInput("day1")
	input_strings := strings.Fields(data)
	integers := make([]int, len(input_strings))
	for ind, str := range input_strings {
		i, _ := strconv.Atoi(str)
		integers[ind] = i
	}
	return integers
}

func Part1() string {
	integers := parseInput()
	increases := 0
	previousDepth := 0
	for ii, depth := range integers {
		if ii == 0 {
			continue
		}
		if depth > previousDepth {
			increases++
		}
		previousDepth = depth
	}
	return fmt.Sprint(increases)
}

func Part2() string {
	integers := parseInput()
	increases := 0
	for ii := 0; ii < len(integers)-4; ii++ {
		currentWindow := integers[ii : ii+3]
		nextWindow := integers[ii+1 : ii+4]
		sumNext := 0
		for _, value := range nextWindow {
			sumNext += value
		}
		sumCurrent := 0
		for _, value := range currentWindow {
			sumCurrent += value
		}
		if sumNext > sumCurrent {
			increases++
		}
	}
	return fmt.Sprint(increases)
}

func All() {
	utils.Profile(Part1)
	utils.Profile(Part2)
}

File: python/puzzles/y2021/day1.py
import utils


def process_input(raw: str):
    return list(map(int, map(str.strip, raw.strip().split("\n"))))


@utils.profile
def part1(raw: str):
    input = process_input(raw)
    increases = 0
    previous_depth = 0
    for ii, depth in enumerate(input[1:]):
        if depth > previous_depth:
            increases += 1
        previous_depth = depth
    return increases


@utils.profile
def part2(raw: str):
    input = process_input(raw)
    increases = 0
    for ii, depth in enumerate(input):
        current_window = input[ii : ii + 3]
        next_window = input[ii + 1 : ii + 4]
        if len(next_window) < 3:
            break  # reached end of list
        if sum(next_window) > sum(current_window):
            increases += 1

    return increases


if __name__ == "__main__":
    raw = utils.load_puzzle_input("2021/day1")
    assert part1(raw) == 1482
    assert part2(raw) == 1518
